Cli: Switch to jopt-simple

rjernst opened the pull request:
We currently use commons-cli for command line parsing, along with complex abstractions built around it. This change simplifies how command line parsing is done, switching to the jopt-simple library.

Some notes:
* bin/elasticsearch no longer supports --foo=bar. It still supports -Dfoo=bar. This is a large simplification.
* Removed the attachments runner cli. This is no longer necessary since it is now in ingest, and can be simulated there.

rmuir commented:
this may still need updated sha1/licensing info etc for the new jar and removal of the old one?

nik9000 commented:
I like it!

nik9000 commented:
But I didn't read it closely enough to admit to "reviewing" it.

rjernst commented:
> this may still need updated sha1/licensing info etc for the new jar and removal of the old one?

Yeah I had only been running unit tests. I will run a full check and fix the shas later tonight.

rjernst commented:
I've fixed sha/license issues. I also changed the -E back to -D for bin/elasticsearch, to minimize changes to integ tests necessary. We still can and should do the change to -E to disambiguate from java's -D, but that can be done as a followup.

uboness commented:
nice cleanup... I'd still consider using `.help` files for the help, at least with common-cli I found that putting it in files gives you much more flexibility on the format/style (specially when it comes to long descriptions and the need to wrap the lines at 80)... similar to man files. That said, I didn't try to run the help with this change to compare.

clintongormley commented:
@rjernst why does it have to be `-E` or `-D` instead of `--`?  This will be a breaking change for every installation of Elasticsearch and, besides, `--` feels more readable to me.

rjernst commented:
> why does it have to be -E or -D instead of --? This will be a breaking change for every installation of Elasticsearch and, besides, -- feels more readable to me.

In order to support --, we have to turn on "leniency" in the option parser, which negates the point of having an option parser to do the option parsing work for us, since we now have to do validation for unknown options ourselves. The way -- was handled in the old parser was crazy. It would take all the unknown options, and try to manually parse them after option parsing was complete, looking for the key=value form, and then adding those to the system properties it sets. We already have 6 other ways to set configuration (and this is not an exaggeration, we have 4 different system properties prefixes, the config file itself, and these two command line alternatives). I understand that it is breaking, but I don't think it is something difficult for a user to update when they see it is no longer supported. As for "readability", I think code complexity outweighs that here, but I also personally don't see any difference.

While I don't personally think it is useful, for the same reason about having 7 different ways to configure settings above, one additional possibility is to still support -- for explicit options we find useful to have. For example, we can add `--path.conf` as a supported option. It is just the arbitrary and dynamic handling of -- that must go.

rjernst commented:
> I'd still consider using .help files for the help, at least with common-cli I found that putting it in files gives you much more flexibility on the format/style (specially when it comes to long descriptions and the need to wrap the lines at 80)... similar to man files

I would not use resource files for help messaging for a number of reasons:
* We have an option parser that we've registered the options with. It seems a waste to not use its ability to generation help!
* The files can easily be out of date and incorrect (which I found numerous times when moving over the help text)
* The tests for them were not useful, they just tested that the file was printed!
* I don't think man file style should be used. If we want man files for eg `bin/elasticsearch`, then we should add those to our distribution, but --help should be usage information.

That said, I do think we can still improve the help output. Right now, we don't have a usage line in the help, but I think that can be added in a followup.

> That said, I didn't try to run the help with this change to compare.

Here is what the help looks like with this PR for `bin/elasticsearch`:
```
[08:54:26][~/code/elasticsearch/distribution/zip/build/distributions/elasticsearch-5.0.0-SNAPSHOT]$ bin/elasticsearch --help
Starts elasticsearch

Option           Description
------           -----------
-D               Configures an Elasticsearch setting
-V, --version    Prints elasticsearch version
                   information and exits
-d, --daemonize  Starts Elasticsearch in the background
-h, --help       show help
-p, --pidfile    Creates a pid file in the specified
                   path on start
-s, --silent     show minimal output
-v, --verbose    show verbose output
```

clintongormley commented:
@rjernst thanks for the explanation

> While I don't personally think it is useful, for the same reason about having 7 different ways to configure settings above, one additional possibility is to still support -- for explicit options we find useful to have. For example, we can add --path.conf as a supported option. It is just the arbitrary and dynamic handling of -- that must go.

No, I'd keep things consistent, otherwise the user will have to look up which options use `--` every time.

jasontedor commented:
> As for "readability", I think code complexity outweighs that here, but I also personally don't see any difference.

I've been through the code (both the old, and the new) by myself and with @rjernst and not having to support multiple ways to do the same thing *greatly* simplifies the code. 

> We already have 6 other ways to set configuration (and this is not an exaggeration, we have 4 different system properties prefixes, the config file itself, and these two command line alternatives).

And I would like to take it even further when I rework #16971 on top of this pull request to remove some of this flexibility as well.

jasontedor commented on core/src/main/java/org/elasticsearch/cli/Terminal.java:
```diff
@@ -117,7 +117,7 @@ public String readText(String prompt) {
     private static class SystemTerminal extends Terminal {
 
-        private final PrintWriter writer = newWriter();
+        private static final PrintWriter writer = newWriter();
```
Since it's static now, change the name to `WRITER`?

jasontedor commented on modules/lang-groovy/build.gradle:
```diff
@@ -38,6 +38,15 @@ thirdPartyAudit.excludes = [
+  'org.apache.commons.cli.OptionBuilder',
+  'org.apache.commons.cli.Options',
+  'org.apache.commons.cli.Parser',
+  'org.apache.commons.cli.PosixParser',
```
I'm just curious why these needed to be added to the excludes?

rjernst replied:
Because groovy depends on commons cli but we don't include it.

nik9000 replied:
Can you add a comment saying that? It'll make whoever touches this six months from now very happy.

jasontedor replied:
Sorry, I get *that* part, I thought that Groovy jarjar'ed the CLI dependency though?

jasontedor replied:
Such as this: https://github.com/apache/groovy/blob/6a9453534ed35eef06a2f481f88776acad3c4af6/gradle/assemble.gradle#L182-L200

rjernst replied:
I ran jdeps against the 2.4.6 indy jar, and there are some references to the jarjar version of commons cli, and some to the real cli package. So it would appear their packaging has problems...

nik9000 replied:
> So it would appear their packaging has problems...

Packaging is hard!

jasontedor replied:
*Sigh*. Thanks for double checking.

jasontedor commented on qa/evil-tests/src/test/java/org/elasticsearch/bootstrap/BootstrapCliParserTests.java:
```diff
@@ -67,195 +64,86 @@ public void clearProperties() {
 
-        ExitStatus status = parser.execute(args("start -Dfoo=bar -Dspam=eggs"));
-        assertStatus(status, OK);
+        execute("-Efoo=bar", "-Espam=eggs");
```
Shouldn't these be `-D` for now?

rjernst replied:
Yeah, I hadn't run the tests since switching back to `-D`. :)

jasontedor commented on qa/evil-tests/src/test/java/org/elasticsearch/bootstrap/BootstrapCliParserTests.java:
```diff
@@ -67,195 +64,86 @@ public void clearProperties() {
+    public void testConfigMalformed() throws Exception {
         UserError e = expectThrows(UserError.class, () -> {
-                parser.parse("start", new String[]{"--foo=bar", "-Dbaz=qux"});
+            execute("-Efoo");
```
Same thing here, `-D`?

jasontedor commented on qa/evil-tests/src/test/java/org/elasticsearch/bootstrap/BootstrapCliParserTests.java:
```diff
@@ -38,18 +26,27 @@
 
 @SuppressForbidden(reason = "modifies system properties intentionally")
-public class BootstrapCliParserTests extends CliToolTestCase {
+public class BootstrapCliParserTests extends CommandTestCase {
```
Can you add a test for an unrecognized option, such as `--network.host`? :wink:

jasontedor commented:
@rjernst This is a *great* change, I left an initial review.

rjernst commented:
@jasontedor @nik9000 I pushed a new commit.

jasontedor commented:
LGTM.

jasontedor commented on modules/lang-groovy/build.gradle:
```diff
@@ -38,6 +38,17 @@ thirdPartyAudit.excludes = [
   'com.thoughtworks.xstream.XStream', 
   'groovyjarjarasm.asm.util.Textifiable', 
+  // commons-cli is referenced by groovy, even though they supposedly
+  // jarjar it. Since we don't use the cli, we don't need the dep.
```
+1

jasontedor commented on qa/evil-tests/src/test/java/org/elasticsearch/bootstrap/BootstrapCliParserTests.java:
```diff
@@ -67,195 +64,93 @@ public void clearProperties() {
-        BootstrapCLIParser parser = new BootstrapCLIParser(terminal);
-        UserError e = expectThrows(UserError.class, () -> {
-                parser.parse("start", new String[]{"--foo=bar", "-Dbaz=qux"});
+    public void testUnknownOption() throws Exception {
```
+1

clintongormley commented:
@rjernst please add a note to breaking changes

jasontedor commented:
@clintongormley I added such a note in #17088 (I carried it over from #16791).

Mpdreamz commented:
Might need to open a new issue for this but 

> $ elasticsearch blahblah 

silently ignores blah blah

which can get hairy if someone forgets an `-E` e.g

> $ elasticsearch -E es.cluster.name="x" es.node.name="y"

the latter setting is ignored.

nik9000 commented:
Yeah, I'd open an issue for that.
